src/merkle/smt/full/proof.rs

```diff
@@ -0,0 +1,80 @@
+use super::{MerklePath, RpoDigest, SmtLeaf, Word};
+
+/// A proof which can be used to assert membership (or non-membership) of key-value pairs in a
+/// [`super::Smt`].
+///
+/// The proof consists of a Merkle path and leaf which describes the node located at the base of the
+/// path.
+#[derive(PartialEq, Eq, Debug)]
+pub struct SmtProof {
+    path: MerklePath,
+    leaf: SmtLeaf,
+}
+
+impl SmtProof {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns a new instance of [`SmtProof`] instantiated from the specified path and leaf.
+    ///
+    /// # Errors
+    /// Returns an error if the path length is not [`SMT_DEPTH`].
+    pub fn new(path: MerklePath, leaf: SmtLeaf) -> Self {
+        Self { path, leaf }
+    }
```

Comment:
I agree. However, we have a similar problem: the `From<(MerklePath, SmtLeaf)> for SmtProof` is also part of the public interface. And we currently need to leave it there because of `SparseMerkleTree::Opening`.

This could be solved by removing the `From<(MerklePath, Self::Leaf)>` trait bound on `SparseMerkleTree`, and instead adding a method `path_and_leaf_to_opening(MerklePath, Self::Leaf) -> Self::Opening`. I will implement this change now since I'm pretty sure this is the way to go.

Reply:
Agreed!

Reply:
Done